src/XAD/Jacobian.hpp

```diff
@@ -0,0 +1,131 @@
+    return matrix;
+}
+
+// adj iterator
+template <class RowIterator, typename T>
+void computeJacobian(const std::vector<xad::AReal<T>> &vec,
+                     std::function<std::vector<xad::AReal<T>>(std::vector<xad::AReal<T>> &)> foo,
+                     RowIterator first, RowIterator last,
+                     xad::Tape<T> *tape = xad::Tape<T>::getActive())
+{
+    auto v(vec);
+    unsigned int domain = static_cast<unsigned int>(vec.size()),
+                 codomain = static_cast<unsigned int>(foo(v).size());
+
+    if (std::distance(first, last) != codomain)
+        throw OutOfRange("Iterator not allocated enough space");
+    static_assert(
+        xad::detail::has_begin<typename std::iterator_traits<RowIterator>::value_type>::value,
+        "RowIterator must dereference to a type that implements a begin() method");
+    std::unique_ptr<xad::Tape<T>> t;
+    if (!tape)
+    {
+        t = std::unique_ptr<xad::Tape<T>>(new xad::Tape<T>());
+        tape = t.get();
+    }
+
+    tape->registerInputs(v);
+
+    auto row = first;
+    for (unsigned int i = 0; i < codomain; i++, row++)
+    {
+        auto col = row->begin();
+        for (unsigned int j = 0; j < domain; j++, col++)
+        {
+            tape->newRecording();
+            xad::AReal<T> y = foo(v)[i];
+            tape->registerOutput(y);
+            derivative(y) = 1.0;
+            tape->computeAdjoints();
+            *col = derivative(v[j]);
+        }
+    }
+}
+
+// fwd 2d vector
+template <typename T>
+std::vector<std::vector<T>> computeJacobian(
+    const std::vector<xad::FReal<T>> &vec,
+    std::function<std::vector<xad::FReal<T>>(std::vector<xad::FReal<T>> &)> foo)
+{
+    auto v(vec);
+    std::vector<std::vector<T>> matrix(foo(v).size(), std::vector<T>(v.size(), 0.0));
+    computeJacobian(vec, foo, begin(matrix), end(matrix));
+    return matrix;
+}
+
+// fwd iterator
+template <class RowIterator, typename T>
+void computeJacobian(const std::vector<xad::FReal<T>> &vec,
+                     std::function<std::vector<xad::FReal<T>>(std::vector<xad::FReal<T>> &)> foo,
+                     RowIterator first, RowIterator last)
+{
+    auto v(vec);
+    unsigned int domain = static_cast<unsigned int>(vec.size()),
+                 codomain = static_cast<unsigned int>(foo(v).size());
+
+    if (std::distance(first, last) != codomain)
+        throw OutOfRange("Iterator not allocated enough space");
+    static_assert(
+        xad::detail::has_begin<typename std::iterator_traits<RowIterator>::value_type>::value,
+        "RowIterator must dereference to a type that implements a begin() method");
+
+    auto row = first;
+    for (unsigned int i = 0; i < codomain; i++, row++)
+    {
+        auto col = row->begin();
+        for (unsigned int j = 0; j < domain; j++, col++)
+        {
+            derivative(v[j]) = 1.0;
+            xad::FReal<T> y = foo(v)[i];
+            *col = derivative(y);
```

Comment:
with forward mode AD, you get the derivative of *all* outputs w.r.t. the one input where you set the derivative to 1.0. There's no need to iterate over the outputs - only over the inputs.

Reply:
Thing is I need to evaluate y after seeding to get the derivative. I came up with this which reduces evaluations for f(x):
```c++
    for (unsigned int i = 0; i < codomain; i++, row++)
    {
        derivative(v[i]) = 1.0;
        auto y = foo(v);

        auto col = row->begin();
        for (unsigned int j = 0; j < domain; j++, col++)
        {
            *col = derivative(y[j]);
            // std::cout << derivative(y[j]) << std::endl;
        }

        derivative(v[i]) = 0.0;
    }
```
Unless there's a function that computes a vector of derivatives, I believe there is no smaller time complexity than this. I believe this computes the Jacobian transposed so I need to look into that.

Reply:
Yes, this is the right way to do it. Cuts the number of function evaluation by `domain` times (which is huge - you can assume that evaluating the function has a high computational cost). 

The transposing might be an issue, yes. Especially if it's not a square matrix (so you can't just transpose after as storage is not sufficient). You may have to store an array of column iterators in an array and assign to each in sequence + move them.